lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,394 @@
+      public FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult> {
+
+  using ValueIdentifier = std::pair<Value, Identifier>;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitDecl;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitExpr;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitStmt;
+
+  void runOnOperation() override;
+
+  // Lowering operations.
+  LogicalResult visitDecl(InstanceOp op);
+  LogicalResult visitExpr(SubfieldOp op);
+  LogicalResult visitStmt(ConnectOp op);
+
+  // Skip over any other operations.
+  LogicalResult visitUnhandledOp(Operation *op) { return success(); }
+  LogicalResult visitInvalidOp(Operation *op) { return success(); }
+
+private:
+  // Lowering module block arguments.
+  void lowerArg(BlockArgument arg, FIRRTLType type);
+
+  // Helpers to manage state.
+  Value addArg(Type type, unsigned oldArgNumber, StringRef nameSuffix = "");
+
+  void setBundleLowering(Value oldValue, StringRef flatField, Value newValue);
+  Value getBundleLowering(Value oldValue, StringRef flatField);
+  void getAllBundleLowerings(Value oldValue, SmallVectorImpl<Value> &results);
+
+  void removeArgLater(unsigned argNumber);
+  void removeOpLater(Operation *op);
+  void cleanup();
+
+  // The builder is set and maintained in the main loop.
+  OpBuilder *builder;
+
+  // State to keep track of arguments and operations to clean up at the end.
+  SmallVector<unsigned, 8> argsToRemove;
+  SmallVector<Operation *, 16> opsToRemove;
+
+  // State to keep a mapping from (Value, Identifier) pairs to flattened values.
+  DenseMap<ValueIdentifier, Value> loweredBundleValues;
+};
+} // end anonymous namespace
+
+/// This is the pass constructor.
+std::unique_ptr<mlir::Pass> circt::firrtl::createLowerFIRRTLTypesPass() {
+  return std::make_unique<FIRRTLTypesLowering>();
+}
+
+// This is the main entrypoint for the lowering pass.
+void FIRRTLTypesLowering::runOnOperation() {
+  auto module = getOperation();
+  auto *body = module.getBodyBlock();
+
+  OpBuilder theBuilder(&getContext());
+  builder = &theBuilder;
+
+  // Lower the module block arguments.
+  SmallVector<BlockArgument, 8> args(body->args_begin(), body->args_end());
+  for (auto arg : args)
+    if (auto type = arg.getType().dyn_cast<FIRRTLType>())
+      lowerArg(arg, type);
+
+  // Lower the operations.
+  for (auto &op : body->getOperations()) {
+    builder->setInsertionPoint(&op);
+    if (failed(dispatchVisitor(&op)))
+      return;
+  }
+
+  cleanup();
+}
+
+//===----------------------------------------------------------------------===//
+// Lowering module block arguments.
+//===----------------------------------------------------------------------===//
+
+// Lower arguments with bundle type by flattening them.
+void FIRRTLTypesLowering::lowerArg(BlockArgument arg, FIRRTLType type) {
+  // Flatten any bundle types.
+  SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+  flattenBundleTypes(type, "", false, fieldTypes);
+
```

Comment:
Good point, I will try to add early exit for those cases.

Reply:
Actually, the reason I wasn't exiting early for the unchanged arguments was to keep the resulting arguments in the same relative order (with bundles flattened of course). I achieve this by copying unchanged arguments to the end of the argument list and rewriting the IR to use the copy.

I think the proper thing to do here is twofold: first, any time we lower a bundle argument, we should just insert the lowered values right after the original bundle in the argument list. Then, we don't need to needlessly copy/rewrite for every unchanged argument.

I will try to take this approach here.

Reply:
And now I remember why this was tricky--as we lower arguments, if we insert a flattened argument into the middle of the argument list, this may interfere with any argument attributes for the original argument that was at that index. I think it may be possible, but the bookkeeping gets more complicated.

So I see three options:
1) copy the unchanged arguments and rewrite the IR accordingly, so we can keep the same relative ordering (what is in the current revision)
2) don't touch unchanged arguments and allow the arguments to change order in the result
3) insert new arguments into the middle of the argument list and figure out the bookkeeping to keep track of argument attributes

I think 2) is not what we want. It would violate the principle of least surprise, at least for me. For example, output ports would often be arguments at the end of the list, and could get shifted to the middle if some inputs were lowered but outputs were not.

It seems like 3) is the best option in the long run since it would avoid some unnecessary work. 

However, I think 1) is sufficient for now. It's doing some unnecessary work in some cases, but the implementation is straightforward and the results are not surprising. I think I'll land this as-is, and if it is worth pursing 3) (or some other approach), I can address it in a follow-up.

Reply:
I landed this pass and opened this issue to track this specific issue for follow up: https://github.com/llvm/circt/issues/247.